### PR TITLE
⚠️ Update controller-runtime to v0.7 ⚠️

### PR DESCRIPTION
**Pull request by vincepri**
Signed-off-by: Vince Prignano <vincepri@vmware.com>

**What this PR does / why we need it**:
Work in progress to prep for v1alpha4, this is unfinished, and everything in here might change, proceed with caution :)

This also requires https://github.com/kubernetes-sigs/cluster-api/pull/3756

/milestone v0.4.0

### CODE REVIEW

**Comment by vincepri** [comment]
🎉

**Comment by vincepri** [comment]
Before merging / having reviews, I'm going to split the changes in separate commits, trying to be mindful that there are lots of changes. Given that most of the changes should go in together, let's keep everything in one PR

**Comment by vincepri** [comment]
/assign @ncdc @sedefsavas @wfernandes @CecileRobertMichon @fabriziopandini 

This is ready for review now, I've split the commits in multiple logical components

**Comment by fabriziopandini** [comment]
This is an amazing work, kudos to @vincepri

**Comment by devigned** [comment]
Is there a `XXXXL` label?

Awesome, @vincepri!

**Comment by ncdc** [comment]
I may have some comments, doing my best to review but getting pulled in multiple directions.
/hold
while I review

File: controllers/remote/stoppable_cache.go
Comment (ncdc) [inline controllers/remote/stoppable_cache.go:35]: Update godoc in this file to no longer reference a stop channel?

File: controllers/cluster_controller.go
Comment (ncdc) [inline controllers/cluster_controller.go:236]: Do you need this type assertion? Isn't child already a client.Object?

Also, we shouldn't need to use meta.Accessor any more?

Comment (vincepri) [reply controllers/cluster_controller.go:236]: `child` is a `metav1.Object` in this case

Comment (vincepri) [reply controllers/cluster_controller.go:236]: I see what you mean, yeah let me change that

**Comment by wfernandes** [comment]
@vincepri Just went through some of the commits. We should open a separate github issue regarding the skipped tests that need to be reworked. It could be separate github issues per test or a single on that tracks them all. WDYT?

File: controllers/cluster_controller.go
Comment (wfernandes) [inline controllers/cluster_controller.go:97]: As per the documentation for `ctrl.LoggerFrom(ctx)`, it seems like it auto fills in the `req.Name` and `req.Namespace` which is great! 

However, it feels like some of the readability is lost because before we had a log with a key `cluster`. Also, it seems we are also removing the `WithName` options when setting up the logger in the reconcilers.
If these logs were aggregated, it may be difficult to read. 
_I'm assuming that the context keys used are generically `name` and `namespace`. I didn't get a chance to dig that deep._

Maybe we can do something like this once in every respective reconciler.
```suggestion
	log := ctrl.LoggerFrom(ctx).WithName("controllers").WithName("Cluster")
```
WDYT?

Comment (ncdc) [reply controllers/cluster_controller.go:97]: I believe https://github.com/kubernetes-sigs/controller-runtime/blob/c000ea850121b53dd769f624f3bfa74a4fcf100f/pkg/controller/controller.go#L129 sets up the controller name that you're looking for, for all controllers.

And yes, the keys are namespace and name: https://github.com/kubernetes-sigs/controller-runtime/blob/c000ea850121b53dd769f624f3bfa74a4fcf100f/pkg/internal/controller/controller.go#L98.

I can see where you're coming from comparing
```
"namespace"="foo" "cluster"="bar"
```
to
```
"namespace"="foo" "name"="bar"
```
but if we also consider the controller name, does it make it any better?
```
"controller"="cluster" "namespace"="foo" "name"="bar"
```
or perhaps this?
```
"controller"="cluster" "reconcile-request"="foo/bar"
```

Comment (vincepri) [reply controllers/cluster_controller.go:97]: Thanks for going through the codebase @ncdc :), there is also https://github.com/kubernetes-sigs/controller-runtime/blob/c000ea850121b53dd769f624f3bfa74a4fcf100f/pkg/builder/controller.go#L253-L257 which sets up the `reconcilerGroup` and `reconcilerKind` in the controller logger

Comment (vincepri) [reply controllers/cluster_controller.go:97]: We should be able to bring some changes upstream if the logs don't fit nicely with our expectations, there is still time until v0.7.0 is released

Comment (wfernandes) [reply controllers/cluster_controller.go:97]: @ncdc @vincepri Thanks for pointing out the name of the controller and reconciler's GVK being added to the logger. I'm not sure where or how our logger is being threaded in, because our current usage is `ctrl.LoggerFrom` which calls https://github.com/kubernetes-sigs/controller-runtime/blob/c000ea850121b53dd769f624f3bfa74a4fcf100f/pkg/log/log.go#L57 which seems to be returning a new logger. 
~_Unless there is some shared logger that is being used within controller-runtime._~
UPDATE: I see that this logger is added into the context. https://github.com/kubernetes-sigs/controller-runtime/blob/c000ea850121b53dd769f624f3bfa74a4fcf100f/pkg/internal/controller/controller.go#L99


I haven't had the chance yet to deploy the branch and view the logs directly myself so it might look fine. This was just something I thought I'd bring to attention but don't believe it's a blocker.

Comment (vincepri) [reply controllers/cluster_controller.go:97]: We should definitely revisit / review the logs after this merges, let's open a different issue? I do want to have consistent logging everywhere as part of the v0.4.0 release, and everything should use the same logger and have consistent names

Comment (wfernandes) [reply controllers/cluster_controller.go:97]: I'm fine with opening a different issue to track this work and ensure that logs are legible and useful to the reader. That seems like a sizable chunk of work anyways.

File: bootstrap/kubeadm/controllers/kubeadmconfig_controller.go
Comment (benmoss) [inline bootstrap/kubeadm/controllers/kubeadmconfig_controller.go:611]: Nit: if we're just gonna panic we can do `c := o.Object.(*clusterv1.Cluster)` and we'll get 

```
panic: interface conversion: client.Object is foo, not *cluster.Cluster
```

File: bootstrap/kubeadm/controllers/kubeadmconfig_controller.go
Comment (benmoss) [inline bootstrap/kubeadm/controllers/kubeadmconfig_controller.go:659]: Panic here too?

File: bootstrap/kubeadm/controllers/kubeadmconfig_controller.go
Comment (benmoss) [inline bootstrap/kubeadm/controllers/kubeadmconfig_controller.go:675]: Panic here too?

File: controllers/cluster_controller.go
Comment (benmoss) [reply controllers/cluster_controller.go:97]: cluster controller logs now look like:

```
controller/cluster "msg"="Reconciliation is paused for this object" "name"="bmo" "namespace"="default" "reconcilerGroup"="cluster.x-k8s.io" "reconcilerKind"="Cluster"
```

**Review by benmoss — APPROVED** [review]
Just some object coercion nits that aren't really a part of this PR anyway 😸 .

File: bootstrap/kubeadm/controllers/kubeadmconfig_controller.go
Comment (vincepri) [reply bootstrap/kubeadm/controllers/kubeadmconfig_controller.go:611]: I'd rather keep the panic explicit

**Review by devigned — APPROVED** [review]
/lgtm

**Pull request merged**
